### ISSUE TITLE
`exception_handler` not triggered with unhandled exception

### ISSUE THREAD

**Issue by mblayman** [Miserlou/Zappa#1353]
## Context

First, thanks for the great project! I really like it. I'm a Rollbar user and I'm trying to integrate Rollbar into my Lambda function by using the `exception_handler` feature. For some reason, the `exception_handler` is not executing even when I deliberately raise an exception in my Lambda function.

## Expected Behavior

I would hope that the exception I raised would at least trigger the custom exception handler function.

## Actual Behavior

The handler function is never called (I have a logging statement in it that is not invoked). Even stranger, it seems like the exception is raised, but "swallowed" by the WSGI layer.

I'm running a Flask app (more accurately, Flask-Ask because this is an Alexa skill) and my route that I'm purposefully erroring looks like this:

```python
@ask.intent('GetSchoolDeadlinesIntent')
def get_school_deadlines(college):
    logging.info('BEFORE THE EXCEPTION')
    raise Exception('Test skill explosion')
```

I narrowed down where I think the problem is to a line in `handler.py`. To test my theory, I added some extra logging to zappa:

```python
                logger.info('IN ZAPPA, before from_app')
                # Execute the application
                response = Response.from_app(self.wsgi_app, environ)
                logger.info('IN ZAPPA, after from_app')
```

When I triggered this code path, I got this from my logs:

```log
[1516111798828] [INFO] 2018-01-16T14:09:58.828Z ef92c013-fac6-11e7-81bc-15ed5fe80f59 IN ZAPPA, before from_app
[1516111798973] [INFO] 2018-01-16T14:09:58.973Z ef92c013-fac6-11e7-81bc-15ed5fe80f59 BEFORE THE EXCEPTION
[1516111798974] [2018-01-16 14:09:58,973] ERROR in app: Exception on / [POST]
[1516111798974] Traceback (most recent call last):
[1516111798974] File "/private/var/folders/5m/h_k9bn516v9cgshtg3v880y40000gn/T/pip-build-MGOmkN/Flask/flask/app.py", line 1982, in wsgi_app
[1516111798974] File "/private/var/folders/5m/h_k9bn516v9cgshtg3v880y40000gn/T/pip-build-MGOmkN/Flask/flask/app.py", line 1614, in full_dispatch_request
[1516111798974] File "/private/var/folders/5m/h_k9bn516v9cgshtg3v880y40000gn/T/pip-build-MGOmkN/Flask/flask/app.py", line 1517, in handle_user_exception
[1516111798974] File "/private/var/folders/5m/h_k9bn516v9cgshtg3v880y40000gn/T/pip-build-MGOmkN/Flask/flask/app.py", line 1612, in full_dispatch_request
[1516111798974] File "/private/var/folders/5m/h_k9bn516v9cgshtg3v880y40000gn/T/pip-build-MGOmkN/Flask/flask/app.py", line 1598, in dispatch_request
[1516111798974] File "/private/var/folders/5m/h_k9bn516v9cgshtg3v880y40000gn/T/pip-build-MGOmkN/flask-ask/flask_ask/core.py", line 670, in _flask_view_func
[1516111798974] File "/var/task/app.py", line 76, in get_school_deadlines
[1516111798974] raise Exception('Test skill explosion')
[1516111798974] Exception: Test skill explosion
[1516111798974] [INFO] 2018-01-16T14:09:58.974Z ef92c013-fac6-11e7-81bc-15ed5fe80f59 IN ZAPPA, after from_app
[1516111798974] [INFO] 2018-01-16T14:09:58.974Z ef92c013-fac6-11e7-81bc-15ed5fe80f59 72.21.217.109 - - [16/Jan/2018:14:09:58 +0000] "POST / HTTP/1.1" 500 291 "" "Apache-HttpClient/UNAVAILABLE (Java/1.8.0_131)" 0/146.435
```

The part about this that I find strange is that it didn't raise the exception to Zappa's handler module. In my read-through of the code, it's critical for the error to raise here if there is any hope of triggering the `exception_handler` function. I'm fairly certain it did not raise because I put logging in the first line of the except clause and I don't see it.

```python
                return zappa_returndict
        except Exception as e:  # pragma: no cover
            logger.info('IN ZAPPA, logging in handler except clause')
```

## Possible Fix
Is there some mode where Werkzeug prints out errors without actually raising them?

## Steps to Reproduce
Hopefully the sample code above is enough to demonstrate where this problem is happening and how you might reproduce it. I can try to supply more if needed.

## Your Environment
* Zappa version used: `0.45.1`
* Operating System and Python version: OS X 10.11.6 and Python **2.7.13**
* The output of `pip freeze`:
```
aniso8601==1.2.0
argcomplete==1.9.2
asn1crypto==0.24.0
base58==0.2.4
blinker==1.4
boto3==1.5.14
botocore==1.8.28
certifi==2017.11.5
cffi==1.11.3
cfn-flip==1.0.0
chardet==3.0.4
click==6.7
configparser==3.5.0
cryptography==2.1.4
docutils==0.14
durationpy==0.5
enum34==1.1.6
flake8==3.5.0
Flask==0.12.1
Flask-Ask==0.9.7
future==0.16.0
futures==3.1.1
hjson==3.0.1
idna==2.6
ipaddress==1.0.19
itsdangerous==0.24
Jinja2==2.10
jmespath==0.9.3
kappa==0.6.0
lambda-packages==0.19.0
MarkupSafe==1.0
mccabe==0.6.1
placebo==0.8.1
pycodestyle==2.3.1
pycparser==2.18
pyflakes==1.6.0
pyOpenSSL==17.0.0
python-dateutil==2.6.1
python-slugify==1.2.4
PyYAML==3.12
requests==2.18.4
rollbar==0.13.17
s3transfer==0.1.12
six==1.11.0
toml==0.9.4
tqdm==4.19.1
troposphere==2.1.2
Unidecode==1.0.22
urllib3==1.22
Werkzeug==0.12
wsgi-request-logger==0.4.6
zappa==0.45.1
```
* Link to your project (optional): not relevant
* Your `zappa_settings.py`: 
```
    "development": {
        "app_function": "app.app",
        "aws_region": "us-east-1",
        "profile_name": "zappa-deploy",
        "project_name": "conductor-skill",
        "runtime": "python2.7",
        "s3_bucket": "zappa-79t7ut9g6",
        "exception_handler": "app.unhandled_exception"
    },
```

Thanks!

**Comment by mblayman** [comment]
Extra data: I tried setting the log level to INFO in the Zappa settings file. This caused the traceback to no longer show in my CloudWatch/zappa logs.

**Comment by mblayman** [comment]
As an additional sanity check, I ran the same kind of test (i.e., raising an exception deliberately) on a regular Flask route to rule out the possibility of the problem being in Flask-Ask. The app behaved in the same way as reported in this issue.

**Comment by mblayman** [comment]
I got a minimal app to reproduce the issue that I'm seeing. This is a Python 2.7 environment. I'm going to run another test on Python 3.6 to see if the behavior is different.

`sample.py`:
```python
import logging

from flask import Flask

app = Flask(__name__)

@app.route('/')
def explode():
    logging.info('3... 2... 1...')
    raise Exception('Boom!')

def unhandled_exception(ex, event, context):
    logging.info('Handling the exception')
    return True
```
Section from `zappa_setting.json`:
```json
    "flasktest": {
        "app_function": "sample.app",
        "aws_region": "us-east-1",
        "profile_name": "zappa-deploy",
        "project_name": "flask-test",
        "runtime": "python2.7",
        "s3_bucket": "zappa-89t7ut9g6",
        "exception_handler": "sample.unhandled_exception"
    },
```
Zappa logs:
```
[1516131516436] [INFO] 2018-01-16T19:38:36.435Z d82e5a24-faf4-11e7-ac61-dfd086039c16 3... 2... 1...
[1516131516436] [2018-01-16 19:38:36,436] ERROR in app: Exception on / [GET]
[1516131516436] Traceback (most recent call last):
[1516131516436] File "/private/var/folders/5m/h_k9bn516v9cgshtg3v880y40000gn/T/pip-build-u0LygK/flask/flask/app.py", line 1982, in wsgi_app
[1516131516436] File "/private/var/folders/5m/h_k9bn516v9cgshtg3v880y40000gn/T/pip-build-u0LygK/flask/flask/app.py", line 1614, in full_dispatch_request
[1516131516436] File "/private/var/folders/5m/h_k9bn516v9cgshtg3v880y40000gn/T/pip-build-u0LygK/flask/flask/app.py", line 1517, in handle_user_exception
[1516131516436] File "/private/var/folders/5m/h_k9bn516v9cgshtg3v880y40000gn/T/pip-build-u0LygK/flask/flask/app.py", line 1612, in full_dispatch_request
[1516131516436] File "/private/var/folders/5m/h_k9bn516v9cgshtg3v880y40000gn/T/pip-build-u0LygK/flask/flask/app.py", line 1598, in dispatch_request
[1516131516436] File "/var/task/sample.py", line 11, in explode
[1516131516436] raise Exception('Boom!')
[1516131516436] Exception: Boom!
[1516131516456] [INFO] 2018-01-16T19:38:36.444Z d82e5a24-faf4-11e7-ac61-dfd086039c16 73.132.125.1 - - [16/Jan/2018:19:38:36 +0000] "GET / HTTP/1.1" 500 291 "" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.11; rv:57.0) Gecko/20100101 Firefox/57.0" 0/8.942
```

**Comment by mblayman** [comment]
Here's the Python 3 version. The end result is still the same. That seems to narrow the problem down to Flask, Zappa, or one of their dependencies.

`zappa_settings.json`:
```
    "flasktest": {
        "app_function": "sample.app",
        "aws_region": "us-east-1",
        "profile_name": "zappa-deploy",
        "project_name": "flask-test-py3",
        "runtime": "python3.6",
        "s3_bucket": "zappa-89t7ut9g6",
        "exception_handler": "sample.unhandled_exception"
    },
```

Logs:
```
[1516132285761] [INFO] 2018-01-16T19:51:25.761Z a1d63c46-faf6-11e7-8eb1-6f3843e1bf68 3... 2... 1...
[1516132285777] [2018-01-16 19:51:25,764] ERROR in app: Exception on / [GET]
[1516132285777] Traceback (most recent call last):
[1516132285777] File "/var/task/flask/app.py", line 1982, in wsgi_app
[1516132285777] response = self.full_dispatch_request()
[1516132285777] File "/var/task/flask/app.py", line 1614, in full_dispatch_request
[1516132285777] rv = self.handle_user_exception(e)
[1516132285777] File "/var/task/flask/app.py", line 1517, in handle_user_exception
[1516132285777] reraise(exc_type, exc_value, tb)
[1516132285777] File "/var/task/flask/_compat.py", line 33, in reraise
[1516132285777] raise value
[1516132285777] File "/var/task/flask/app.py", line 1612, in full_dispatch_request
[1516132285777] rv = self.dispatch_request()
[1516132285777] File "/var/task/flask/app.py", line 1598, in dispatch_request
[1516132285777] return self.view_functions[rule.endpoint](**req.view_args)
[1516132285777] File "/var/task/sample.py", line 11, in explode
[1516132285777] raise Exception('Boom!')
[1516132285777] Exception: Boom!
[1516132285777] [INFO] 2018-01-16T19:51:25.777Z a1d63c46-faf6-11e7-8eb1-6f3843e1bf68 73.132.125.1 - - [16/Jan/2018:19:51:25 +0000] "GET / HTTP/1.1" 500 291 "" "python-requests/2.18.4" 0/38.68
```

**Comment by mblayman** [comment]
I realized one of the problems I'm having. I didn't have `app.config['PROPAGATE_EXCEPTIONS'] = True`. :man_facepalming: 

I'm still having a problem though. I *think* there might be too many `except Exception` blocks in here.

`LambdaHandler.lambda_handler` only calls the `exception_handler` if it gets an exception:

```python
    def lambda_handler(cls, event, context):  # pragma: no cover
        handler = cls()
        exception_handler = handler.settings.EXCEPTION_HANDLER
        try:
            return handler.handler(event, context)
        except Exception as ex:
            exception_processed = cls._process_exception(exception_handler=exception_handler,
                                                         event=event, context=context, exception=ex)
```

This is totally reasonable. Unfortunately the "Normal web app flow" in the `handler` method **also** has an `except Exception` block that does not re-throw anything (instead, it generates a 500 response). I believe this means that it is pretty unlikely for the `exception_handler` to be triggered.

So, is `exception_handler` *not* appropriate for unhandled exceptions? I'm thinking I might need to find a facility in Flask itself to do my Rollbar collection. I hope I'm mistaken or misunderstand some part of this.

**Comment by efimerdlerkravitz** [comment]
As a side note I'm using Django and Sentry and it works as expected. Pay attention that the signature of handle_error is valid, e.g. `def catch_exception(e, event, context):`

**Comment by mszaro** [comment]
@mblayman, have you made any progress here? I'm in a similar situation as you - Python 3 app (Flask), trying to get Rollbar working on Lambda. Doesn't seem to pick up any of the exceptions.

**Comment by mblayman** [comment]
I ended up handling exceptions in the Flask layer instead of Zappa. You can see what I did here: https://github.com/mblayman/conductor-skill/blob/master/app.py

Hope that helps!

**Comment by jneves** [comment]
Ok, I just tripped on this one and now understand the issue: handler catches all the exceptions for web requests to create the 500 response, preventing them from being propagated to lambda_handler (which is the one calling the exception_handler).

**Comment by scoates** [comment]
Fixed in #1416

**Comment by mblayman** [comment]
Thanks all!
